### ISSUE TITLE
sql,*: tracking issue for sql overheads

### ISSUE THREAD

**Issue by ajwerner** [cockroachdb/cockroach#88777]
**Describe the problem**

[Here](https://github.com/cockroachdb/cockroach/files/9650255/profile.14.pb.gz) is a profile
captured over a single-node run of kv95. The profile captures over 15 minutes of samples. This profile contains no networking for distributed KV execution,
and, thus is not a great representation of the overheads our network stack adds. Nevertheless, it provides
an interesting lens to look examine the cockroach process. 

Note also that KV95 represents just about the worst case in terms of amortizing overheads. Many of the things which show up here would pale in comparison to planning and execution overheads of more complex queries. Nevertheless, it's valuable to understand the sources of overheads in simple workloads. 

Note that the syscalls represent something like 5% of the CPU time.
| category | percent |
|-------| -----|
| syscall: write | 4.55% |
| syscall: read | 0.86% |
| VFS.sync | 0.1% |

kvserver grabs about 22.17% of the CPU time.

Below find a list of interesting SQL stack traces that add up to around 35% of all of the CPU time.

* 7.97 [`(*sql.connExecutor).transitionsStateApplyWrapper`](https://github.com/cockroachdb/cockroach/blob/2b799292b0945cd4502d1f127a4522b67772746a/pkg/sql/conn_executor.go#L2102)
    * This is varied: almost 2% in stats things, 0.8% releasing leases, 1.45% making a new KV transaction
<img width="1657" alt="Screen Shot 2022-09-26 at 7 56 11 PM" src="https://user-images.githubusercontent.com/1839234/192400951-8d55782c-b4e1-45ad-9505-d2f5180d12a5.png">

* 7.7% [`(*memo.Memo).IsStale`](https://github.com/cockroachdb/cockroach/blob/c713cffddffbb2718ebb39d84d38d281b0ed21ca/pkg/sql/plan_opt.go#L509-L511)

	* Most of this is in catalog code -- the use of btrees as opposed to hash maps seems to hurt.
	* <img width="1665" alt="Screen Shot 2022-09-26 at 7 40 07 PM" src="https://user-images.githubusercontent.com/1839234/192399339-928fc70e-94c8-49d6-8009-b895ca832518.png">
* 5.43% [`(*sql.DistSQLPhysicalPlanner).setupFlows`](https://github.com/cockroachdb/cockroach/blob/2675c7c0481bb33eb9f43f1c2b52c152fca3ae2d/pkg/sql/distsql_running.go#L653-L655)
     <img width="1657" alt="Screen Shot 2022-09-26 at 7 45 58 PM" src="https://user-images.githubusercontent.com/1839234/192399908-f57bfa71-44df-448a-b0c0-ae622e7930cd.png">

* 4.34% [`(*sql.planner).makeOptimizerPlan`](https://github.com/cockroachdb/cockroach/blob/dab432a7258c2b8ad2c713b269c32da2366aa012/pkg/sql/conn_executor_exec.go#L1432)
	 <img width="1657" alt="Screen Shot 2022-09-26 at 7 40 55 PM" src="https://user-images.githubusercontent.com/1839234/192399438-50f8a16d-1a9c-4910-a29f-e4c1cc69795a.png">
* 2.32% [`(*sql.connExecutor).recordStatementSummary`](https://github.com/cockroachdb/cockroach/blob/dab432a7258c2b8ad2c713b269c32da2366aa012/pkg/sql/conn_executor_exec.go#L1218-L1221)

<img width="1657" alt="Screen Shot 2022-09-26 at 7 58 58 PM" src="https://user-images.githubusercontent.com/1839234/192401185-3b3966d0-e0ea-42c7-8235-2590a6233263.png">

* 1.83% [`(*sql.connExecutor).addActiveQuery`](https://github.com/cockroachdb/cockroach/blob/dab432a7258c2b8ad2c713b269c32da2366aa012/pkg/sql/conn_executor_exec.go#L313)
    * @jordanlewis is doing something about this (I think) in https://github.com/cockroachdb/cockroach/pull/86968
<img width="1657" alt="Screen Shot 2022-09-26 at 8 10 04 PM" src="https://user-images.githubusercontent.com/1839234/192402269-ecda35d3-f776-4600-850e-8f9652f80ea5.png">

* 1.72% [(*colfetcher.cFetcher).resetBatch](https://github.com/cockroachdb/cockroach/blob/02b028a12b1ac69f4cc9f86c00e66b3b8f4f8c9b/pkg/sql/colfetcher/cfetcher.go#L663)
<img width="1657" alt="Screen Shot 2022-09-26 at 8 03 53 PM" src="https://user-images.githubusercontent.com/1839234/192401760-e173b160-749d-44c1-b6e0-bb6bed872a62.png">

* 1.68% [`(*sql.connExecutor).updateTxnRewindPosMaybe`](https://github.com/cockroachdb/cockroach/blob/2b799292b0945cd4502d1f127a4522b67772746a/pkg/sql/conn_executor.go#L2190)
<img width="1790" alt="Screen Shot 2022-09-27 at 9 01 34 AM" src="https://user-images.githubusercontent.com/1839234/192534583-a723debf-67a3-4f39-b6fd-317ddb72773b.png">

* 1.27% [`(*sql.DistSQLPhysicalPlanner).createPhysPlan`](https://github.com/cockroachdb/cockroach/blob/2675c7c0481bb33eb9f43f1c2b52c152fca3ae2d/pkg/sql/distsql_running.go#L1598)
     <img width="1657" alt="Screen Shot 2022-09-26 at 7 42 46 PM" src="https://user-images.githubusercontent.com/1839234/192399751-72fa804b-e484-44c6-a08e-17af3aa915c0.png">

* 1.25%  return from [`(*sql.DistSQLPlanner).PlanAndRunAll`](https://github.com/cockroachdb/cockroach/blob/2675c7c0481bb33eb9f43f1c2b52c152fca3ae2d/pkg/sql/distsql_running.go#L1348)
     <img width="1657" alt="Screen Shot 2022-09-26 at 7 50 20 PM" src="https://user-images.githubusercontent.com/1839234/192400396-098fc496-8755-4e4f-a1aa-a4d4fbe100b7.png">

 * 1.01% [`(*sql.instrumentationHelper).SetIndexRecommendations`](https://github.com/cockroachdb/cockroach/blob/dab432a7258c2b8ad2c713b269c32da2366aa012/pkg/sql/conn_executor_exec.go#L1213)
      <img width="1657" alt="Screen Shot 2022-09-26 at 7 48 21 PM" src="https://user-images.githubusercontent.com/1839234/192400110-67ec2797-1648-4cc6-9649-b06fe9bc32fd.png">

**Additional context**

Note that this does not explore other overheads like grpc, because we don't send any requests in this single-node workload. 

Jira issue: CRDB-19988

**Comment by rytaft** [comment]
Might also be useful: https://cockroachlabs.slack.com/archives/C03NNC21L31/p1660193368559309

**Comment by yuzefovich** [comment]
@mgartner do you think we should close this issue given that we're addressing this via perf efficiency efforts (and have already addressed some of the things called out)?

**Comment by mgartner** [comment]
Yes, let's close this. Note that not all have been addressed. But I don't think they are very hard to find in a profile and there's no sense in keeping an issue open unless we know we can address the overhead.
